tests/integration.rs

```diff
@@ -4356,6 +4356,97 @@ mod tests {
+
+            let info: serde_json::Value = serde_json::from_slice(&cmd_output).unwrap_or_default();
+            let total_mem = &info["config"]["memory"]["size"]
+                .to_string()
+                .parse::<u64>()
+                .unwrap();
+            let actual_mem = &info["memory_actual_size"]
+                .to_string()
+                .parse::<u64>()
+                .unwrap();
+            total_mem - actual_mem
+        }
+
+        #[test]
+        #[cfg(target_arch = "x86_64")]
+        fn test_virtio_balloon() {
+            let focal = UbuntuDiskConfig::new(FOCAL_IMAGE_NAME.to_string());
+            let guest = Guest::new(Box::new(focal));
+            let mut workload_path = dirs::home_dir().unwrap();
+            workload_path.push("workloads");
+
+            let kernel_path = direct_kernel_boot_path();
+
+            let api_socket = temp_api_path(&guest.tmp_dir);
+
+            let guest_memory_size_gb = 32;
+
+            //Let's start a 32G guest with balloon occupied 10G memory
+            let mut child = GuestCommand::new(&guest)
+                .args(&["--api-socket", &api_socket])
+                .args(&["--cpus", "boot=4"])
+                .args(&[
+                    "--memory",
+                    format!("size={}G", guest_memory_size_gb).as_str(),
+                ])
+                .args(&["--kernel", kernel_path.to_str().unwrap()])
+                .args(&["--cmdline", DIRECT_KERNEL_BOOT_CMDLINE])
+                .args(&["--balloon", "size=10737418240,deflate_on_oom=true"])
+                .default_disks()
+                .default_net()
+                .capture_output()
+                .spawn()
+                .unwrap();
+
+            let r = std::panic::catch_unwind(|| {
+                guest.wait_vm_boot(None).unwrap();
+
+                // Wait guest boots up, and then check balloon_mem
+                thread::sleep(std::time::Duration::new(10, 0));
+
+                let orig_balloon = get_balloon_mem(&api_socket);
+                println!("The original balloon memory size is {}Byte", orig_balloon);
+                assert!(orig_balloon > 0);
+                assert!(orig_balloon == 10737418240);
+
+                // Two steps to verify if the 'deflate_on_oom' parameter works.
+                // 1st: run a command in guest to eat up memory heavily, which
+                // will consume much more memory than $(total_mem - balloon_mem)
+                // to trigger an oom.
+                let guest_ip = guest.network.guest_ip.clone();
+                thread::spawn(move || {
+                    ssh_command_ip(
+                        "sudo stress --vm 25 --vm-keep --vm-bytes 1G --timeout 50",
+                        &guest_ip,
+                        DEFAULT_SSH_RETRIES,
+                        DEFAULT_SSH_TIMEOUT,
+                    )
+                    .unwrap();
+                })
+                .join()
+                .unwrap();
```

Comment:
Don't use the timeout, otherwise stress will stop after 50s, and don't use `join()`, otherwise you're going to wait for the stress command to finish. The point is to run the stress command in the background, and check for the balloon size in parallel. So here is my suggestion:
```suggestion
                thread::spawn(move || {
                    ssh_command_ip(
                        "sudo stress --vm 25 --vm-keep --vm-bytes 1G",
                        &guest_ip,
                        DEFAULT_SSH_RETRIES,
                        DEFAULT_SSH_TIMEOUT,
                    )
                    .unwrap();
                });
                
                // Wait for the balloon to deflate
                thread::sleep(std::time::Duration::new(10, 0));
```

Reply:
> Don't use the timeout, otherwise stress will stop after 50s, and don't use `join()`, otherwise you're going to wait for the stress command to finish. The point is to run the stress command in the background, and check for the balloon size in parallel. So here is my suggestion:

Sorry for my previous wrong explanation. :( Actually, after `stress`, if an oom is triggered, the balloon memory will not restore to its original size; but if no oom is triggered, the `free -h`'s result will restore. The previous failure reason is the timeout too short to trigger an oom.     I will try to use smaller sizes so that wait less time and make sure an oom is indeed triggered to validate the deflate_on_oom feature. And maybe run in order without using thread::spawn().